src/libcore/convert.rs

```diff
@@ -0,0 +1,146 @@
+//! conversions from one type to another. They follow the standard
+//! Rust conventions of `as`/`to`/`into`/`from`.
+
+#![unstable(feature = "convert",
+            reason = "recently added, experimental traits")]
+
+use option::Option;
+use result::Result;
+use marker::Sized;
+
+/// A cheap, reference-to-reference conversion.
+pub trait As<T: ?Sized> {
+    /// Perform the conversion.
+    fn convert_as(&self) -> &T;
+}
+
+/// A cheap, mutable reference-to-mutable reference conversion.
+pub trait AsMut<T: ?Sized> {
+    /// Perform the conversion.
+    fn convert_as_mut(&mut self) -> &mut T;
+}
+
+/// A possibly expensive conversion from a reference.
+pub trait To<T> {
+    /// Perform the conversion.
+    fn convert_to(&self) -> T;
+}
+
+/// A conversion that consumes `self`, which may or may not be
+/// expensive.
+pub trait Into<T>: Sized {
+    /// Perform the conversion.
+    fn convert_into(self) -> T;
+}
+
+/// Construct `Self` via a conversion.
+pub trait From<T> {
+    /// Perform the conversion.
+    fn from(T) -> Self;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// GENERIC IMPLS
+////////////////////////////////////////////////////////////////////////////////
+
+// As implies To
+impl<'a, T: ?Sized, U: ?Sized> To<&'a U> for &'a T where T: As<U> {
+    fn convert_to(&self) -> &'a U {
+        (*self).convert_as()
+    }
+}
+
+// As lifts over &
+impl<'a, T: ?Sized, U: ?Sized> As<U> for &'a T where T: As<U> {
+    fn convert_as(&self) -> &U {
+        <T as As<U>>::convert_as(*self)
+    }
+}
+
+// As lifts over &mut
+impl<'a, T: ?Sized, U: ?Sized> As<U> for &'a mut T where T: As<U> {
+    fn convert_as(&self) -> &U {
+        <T as As<U>>::convert_as(*self)
+    }
+}
+
+// AsMut implies Into
+impl<'a, T: ?Sized, U: ?Sized> Into<&'a mut U> for &'a mut T where T: AsMut<U> {
+    fn convert_into(self) -> &'a mut U {
+        (*self).convert_as_mut()
+    }
+}
+
+// AsMut lifts over &mut
+impl<'a, T: ?Sized, U: ?Sized> AsMut<U> for &'a mut T where T: AsMut<U> {
+    fn convert_as_mut(&mut self) -> &mut U {
+        (*self).convert_as_mut()
+    }
+}
+
+// To implies Into
+impl<'a, T: ?Sized, U> Into<U> for &'a T where T: To<U> {
+    fn convert_into(self) -> U {
+        self.convert_to()
+    }
+}
+
+// Into implies From
+impl<T, U> From<T> for U where T: Into<U> {
+    fn from(t: T) -> U {
+        t.convert_into()
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// CONCRETE IMPLS
+////////////////////////////////////////////////////////////////////////////////
+
+impl<T> As<[T]> for [T] {
+    fn convert_as(&self) -> &[T] {
+        self
+    }
+}
+
+impl<T> AsMut<[T]> for [T] {
+    fn convert_as_mut(&mut self) -> &mut [T] {
+        self
+    }
+}
+
+impl As<str> for str {
+    fn convert_as(&self) -> &str {
+        self
+    }
+}
+
+impl<T> As<[T]> for Option<T> {
+    fn convert_as(&self) -> &[T] {
+        self.as_slice()
+    }
+}
```

Comment:
My fear in marking these unstable is that there's no warning on usage of an unstable *impl*, only on an unstable trait, and it's not clear that we'll get it fixed for 1.0.